API Docs (#37)
sourcebots/sbot: docs/source/api/arduino.rst

Arduino API
===========

The `Arduino <https://store.arduino.cc/arduino-uno-rev3>`__
provides a total of 18 pins for either digital input or output (labelled
2 to 13 and A0 to A5), including 6 for analogue input (labelled A0 to
A5).

Accessing the Arduino
---------------------

The Arduino can be accessed using the ``arduino`` property of
the ``Robot`` object.

.. code:: python

   my_arduino = r.arduino

You can use the GPIO *(General Purpose Input/Output)* pins for anything,
from microswitches to LEDs. GPIO is only available on pins 2 to 13 and A0 to A5
because pins 0 and 1 are reserved for communication with the rest of our
kit.

> Setting the pin mode

Reviewer: This is still not in the right place (it's before the introduction to the concept of pin modes - the section entitled "Pin modes" below). Due to time constraints I'll just make a PR to show how I think this should look.

Reviewer: --> #42

Arduino API
===========

The `Arduino <https://store.arduino.cc/arduino-uno-rev3>`__
provides a total of 18 pins for either digital input or output (labelled
2 to 13 and A0 to A5), including 6 for analogue input (labelled A0 to
A5).

Accessing the Arduino
---------------------

The Arduino can be accessed using the ``arduino`` property of
the ``Robot`` object.

.. code:: python

   my_arduino = r.arduino

You can use the GPIO *(General Purpose Input/Output)* pins for anything,

> from microswitches to LEDs. GPIO is only available on pins 2 to 12, and A0 to A5

Reviewer: Removing the comma because the sentence flows weirdly when it's left there. Including pin 13 for consistency with the rest of the text.

```suggestion
from microswitches to LEDs. GPIO is only available on pins 2 to 13 and A0 to A5
```

.. Hint:: The input modes closely resemble those of an
   Arduino. More information on them can be found in `their
   docs <https://www.arduino.cc/en/Tutorial/DigitalPins>`__.

``GPIOPinMode.DIGITAL_INPUT``
~~~~~~~~~~~~~~~~~

In this mode, the digital state of the pin (whether it is high or low)
can be read.

.. code:: python
   
   pin_value = r.arduinos.pins[4].digital_state

``GPIOPinMode.DIGITAL_INPUT_PULLUP``
~~~~~~~~~~~~~~~~~~~~~~~~

Same as ``GPIOPinMode.DIGITAL_INPUT``, but with an internal `pull-up
resistor <https://learn.sparkfun.com/tutorials/pull-up-resistors>`__
enabled.

.. code:: python
   
   pin_value = r.arduinos.pins[4].digital_state

``GPIOPinMode.DIGITAL_OUTPUT``
~~~~~~~~~~~~~~~~~~~~~~~

In this mode, we can set binary values of ``0V`` or ``5V`` to the pin.

.. code:: python
   
   r.arduinos.pins[4].digital_state = True
   r.arduinos.pins[6].digital_state = False

``GPIOPinMode.ANALOGUE_INPUT``
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

In this mode, we can read the analogue voltage of the pin.

Only pins A0 - A5 support this.

.. code:: python
   
    from sbot import AnaloguePin

   pin_value = r.arduinos.pins[AnaloguePin.A0].analogue_value

> Setting the pin mode

Reviewer: I think this section is better placed before the descriptions of each mode, to match the order in which the user should put the described operations in their code.

--------------------

.. code:: python

   from robot import PinMode

   r.arduino.pins[3].mode = PinMode.INPUT_PULLUP

Analogue pins
-------------

Certain sensors output analogue signals rather than digital ones, and so

> have to be read differently. The servo assembly has six analogue inputs,

Reviewer: "servo assembly" is now a legacy term.

```suggestion
have to be read differently. The arduino has six analogue inputs,
```

> of which four (analogue pins 4 and 5) are reserved for communication

Reviewer: This phrase is no longer needed since we no longer use a servo shield.

between the Arduino and the servo shield.

.. Hint:: Analogue signals can have any voltage, while digital
signals can only take on one of two voltages. You can read more about
digital vs analogue signals `here <https://learn.sparkfun.com/tutorials/analog-vs-digital>`__.

> To read the value of an analogue pin, call ``read_analogue()`` on the

Reviewer: This paragraph and its accompanying code sample need to be updated to the `sbot` API.

servo board. This will give you the value of all analogue pins.

.. code:: python

   r.servo_board.read_analogue()["a0"]
   >>> 1.569

.. Hint:: The values are the voltages read on the pins,
   between 0 and 5.

Ultrasound Sensors
------------------

You can also measure distance using an ultrasound sensor from the arduino.

.. code:: python
   

>    # TX: 4

Reviewer: Use correct pin naming.

```suggestion
   # Trigger pin: 4
```

>    # RX: 5

Reviewer: Use correct pin naming.

```suggestion
   # Echo pin: 5
```

>    u = r.get_ultrasound(4, 5)

Reviewer: I think this makes more sense:

```suggestion
   u = r.arduino.get_ultrasound(4, 5)
```

There isn't actually a correct answer to this at the moment since `get_ultrasound` is not defined on either the `Robot` class or the `SBArduinoBoard` class at the moment (now tracked as part of j5api/j5#66).

Reviewer: Since #40, the syntax is now:

```suggestion
   u = r.arduino.ultrasound_sensors[4, 5]
```
